internal/pkg/manifest/validate.go

```diff
@@ -2232,7 +2234,7 @@ func (i ImageLocationOrBuild) validate() error {
 	if err := i.Build.validate(); err != nil {
 		return fmt.Errorf(`validate "build": %w`, err)
 	}
-	if !i.Build.isEmpty() && i.Location != nil {
+	if i.Build.isEmpty() == (i.Location == nil) {
```

Comment:
Won't this error out if we have ⬇️  ?
```
sidecars:
  nginx:
     image: <mylocation>
```

Do you know what happens if we had ⬇️ 
```
name: mysvc

image:
   port: 80
```
That is, neither `location` or `build` is set for the _main container_. Do we error out on validation, or do we error out when we try to build the image?

Reply:
> In case of below config ⬇️ for sidecars
> ```
> sidecars:
>  nginx:
>     image: <mylocation>
> ```
> we do not error out in this case.

I agree that we shouldn't! 


Correct me if I'm wrong though, with current logic (the revision that i'm looking at rn), we will:
1. Go to `if s.Image.IsZero() {` (L1495); because Image is not zero (it has the string) we don't error out
2. Go to `if err := s.Image.validate(); err != nil {`, where `s.image` is an `ImageLocationOrBuild`), this goes to the code that this comment is placed under. It will error out with `errFieldMutualExclusive` because both `i.Build.isEmpty`  and `i.Location == nil` are true.

Reply:
exactly this is true!
`if err := s.Image.validate(); err != nil ` will call the validation of `Union` type over here which will validate both Basic and Advanced Type.
https://github.com/aws/copilot-cli/blob/mainline/internal/pkg/manifest/union.go#L145

In this case we have `Image  Union[*string, ImageLocationOrBuild]`
as there is no validation on the `Basic `type which is just a string, but we have validation on `ImageLocationOrBuild`

Reply:
@Lou1415926 I _think_ that `SidecarConfig.Image` has a couple layers of nesting that is making this confusing:
- `SidecarConfig.Image` is the union type between `*string` and `ImageLocationOrBuild`  
- `SidecarConfig.Image.Advanced` is how to access the `ImageLocationOrBuild` field
- `SidecarConfig.Image.validate()` calls `ImageLocationOrBuild.validate()` only if the config set `build` or `location`

Therefore, in the situation you proposed:
```
sidecars:
  nginx:
    image: <mylocation>
```

My understanding is that `s.Image.IsZero()` returns false, as you said. Then, `s.Image.Validate()` will _**not call any validation function**_, because the underlying value of the union is `*string`, which doesn't have a `validate()` function on it!

Reply:
makes sense!!

Reply:
ah @KollaAdithya you beat me too it! I agree with your synopsis 👍